// @azure/msal-common/dist/error/ClientAuthErrorCodes.js
/*! msal-node-wrapper v1.0.0-beta 2024-02-02 */
'use strict';
'use strict';

/*! @azure/msal-common v14.6.1 2024-01-23 */
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
const clientInfoDecodingError = "client_info_decoding_error";
const clientInfoEmptyError = "client_info_empty_error";
const tokenParsingError = "token_parsing_error";
const nullOrEmptyToken = "null_or_empty_token";
const endpointResolutionError = "endpoints_resolution_error";
const networkError = "network_error";
const openIdConfigError = "openid_config_error";
const hashNotDeserialized = "hash_not_deserialized";
const invalidState = "invalid_state";
const stateMismatch = "state_mismatch";
const stateNotFound = "state_not_found";
const nonceMismatch = "nonce_mismatch";
const authTimeNotFound = "auth_time_not_found";
const maxAgeTranspired = "max_age_transpired";
const multipleMatchingTokens = "multiple_matching_tokens";
const multipleMatchingAccounts = "multiple_matching_accounts";
const multipleMatchingAppMetadata = "multiple_matching_appMetadata";
const requestCannotBeMade = "request_cannot_be_made";
const cannotRemoveEmptyScope = "cannot_remove_empty_scope";
const cannotAppendScopeSet = "cannot_append_scopeset";
const emptyInputScopeSet = "empty_input_scopeset";
const deviceCodePollingCancelled = "device_code_polling_cancelled";
const deviceCodeExpired = "device_code_expired";
const deviceCodeUnknownError = "device_code_unknown_error";
const noAccountInSilentRequest = "no_account_in_silent_request";
const invalidCacheRecord = "invalid_cache_record";
const invalidCacheEnvironment = "invalid_cache_environment";
const noAccountFound = "no_account_found";
const noCryptoObject = "no_crypto_object";
const unexpectedCredentialType = "unexpected_credential_type";
const invalidAssertion = "invalid_assertion";
const invalidClientCredential = "invalid_client_credential";
const tokenRefreshRequired = "token_refresh_required";
const userTimeoutReached = "user_timeout_reached";
const tokenClaimsCnfRequiredForSignedJwt = "token_claims_cnf_required_for_signedjwt";
const authorizationCodeMissingFromServerResponse = "authorization_code_missing_from_server_response";
const bindingKeyNotRemoved = "binding_key_not_removed";
const endSessionEndpointNotSupported = "end_session_endpoint_not_supported";
const keyIdMissing = "key_id_missing";
const noNetworkConnectivity = "no_network_connectivity";
const userCanceled = "user_canceled";
const missingTenantIdError = "missing_tenant_id_error";
const methodNotImplemented = "method_not_implemented";
const nestedAppAuthBridgeDisabled = "nested_app_auth_bridge_disabled";

exports.authTimeNotFound = authTimeNotFound;
exports.authorizationCodeMissingFromServerResponse = authorizationCodeMissingFromServerResponse;
exports.bindingKeyNotRemoved = bindingKeyNotRemoved;
exports.cannotAppendScopeSet = cannotAppendScopeSet;
exports.cannotRemoveEmptyScope = cannotRemoveEmptyScope;
exports.clientInfoDecodingError = clientInfoDecodingError;
exports.clientInfoEmptyError = clientInfoEmptyError;
exports.deviceCodeExpired = deviceCodeExpired;
exports.deviceCodePollingCancelled = deviceCodePollingCancelled;
exports.deviceCodeUnknownError = deviceCodeUnknownError;
exports.emptyInputScopeSet = emptyInputScopeSet;
exports.endSessionEndpointNotSupported = endSessionEndpointNotSupported;
exports.endpointResolutionError = endpointResolutionError;
exports.hashNotDeserialized = hashNotDeserialized;
exports.invalidAssertion = invalidAssertion;
exports.invalidCacheEnvironment = invalidCacheEnvironment;
exports.invalidCacheRecord = invalidCacheRecord;
exports.invalidClientCredential = invalidClientCredential;
exports.invalidState = invalidState;
exports.keyIdMissing = keyIdMissing;
exports.maxAgeTranspired = maxAgeTranspired;
exports.methodNotImplemented = methodNotImplemented;
exports.missingTenantIdError = missingTenantIdError;
exports.multipleMatchingAccounts = multipleMatchingAccounts;
exports.multipleMatchingAppMetadata = multipleMatchingAppMetadata;
exports.multipleMatchingTokens = multipleMatchingTokens;
exports.nestedAppAuthBridgeDisabled = nestedAppAuthBridgeDisabled;
exports.networkError = networkError;
exports.noAccountFound = noAccountFound;
exports.noAccountInSilentRequest = noAccountInSilentRequest;
exports.noCryptoObject = noCryptoObject;
exports.noNetworkConnectivity = noNetworkConnectivity;
exports.nonceMismatch = nonceMismatch;
exports.nullOrEmptyToken = nullOrEmptyToken;
exports.openIdConfigError = openIdConfigError;
exports.requestCannotBeMade = requestCannotBeMade;
exports.stateMismatch = stateMismatch;
exports.stateNotFound = stateNotFound;
exports.tokenClaimsCnfRequiredForSignedJwt = tokenClaimsCnfRequiredForSignedJwt;
exports.tokenParsingError = tokenParsingError;
exports.tokenRefreshRequired = tokenRefreshRequired;
exports.unexpectedCredentialType = unexpectedCredentialType;
exports.userCanceled = userCanceled;
exports.userTimeoutReached = userTimeoutReached;
